src/clients/FtpGateway.cc

```diff
@@ -704,18 +702,22 @@ ftpListParseParts(const char *buf, struct Ftp::GatewayFlags flags)
                 p->size = atoi(ct + 1);
                 break;
 
-            case 'm':
-                tm = (time_t) strtol(ct + 1, &tmp, 0);
-
-                if (tmp != ct + 1)
+            case 'm': {
+                char *tmp;
+                const auto tm = (time_t) strtol(ct + 1, &tmp, 0);
+                if (tmp == ct + 1)
                     break;  /* not a valid integer */
 
+                const char* cts = ctime(&tm);
+                if (!cts)
+                    break;
+
                 safe_free(p->date); // TODO: properly handle multiple p->name occurrences
-                p->date = xstrdup(ctime(&tm));
 
-                *(strstr(p->date, "\n")) = '\0';
+                p->date = xstrndup(cts, strcspn(cts, "\n"));
 
                 break;
+            }
```

Comment:
This code should be like so:
```c++
case 'm': {
    char *tmp;
    const auto tm = (time_t) strtol(ct + 1, &tmp, 0);
    if (tmp == ct + 1)
        break;  /* not a valid integer */
    if (const auto cts = std::ctime(&tm)) {
        xfree(p->date); // TODO: properly handle multiple p->name occurrences
        p->date = xstrndup(cts, strcspn(cts, "\n"));
    }
}
break;
```

Note the final `break` is outside the brackets. Looks funny, but ensures we do not confuse the scoping here as intending to be a fallthrough case.

Reply:
FWIW, I disagree with that `break` placement recommendation (but will not debate that here). Authors choice.